### ISSUE TITLE
Incorrect route rendered when conditionally rendering Outlet in Route with same path as sibling Route

### ISSUE THREAD

**Issue by toonddd** [solidjs/solid-app-router#123]
There is an oddity when conditionally rendering an Outlet on a Route, when that route has the same path as another Route at the same nesting level. The wrong route is shown and even when rendering said Outlet again, the app routing stays in a broken state until refreshed.

Steps to reproduce:

1. run the app code below.
2. go to route 1 by adding /1 as path in the browser's address bar.
3. switch between route /1 and /2 by clicking the links, observe it works as intended.
4. from route /2, click the "trigger bug" button.
5. observe fallback content from route /1 gets rendered, instead of the content from route /2

Code to reproduce:
```
import { NavLink, Outlet, Route, Routes } from 'solid-app-router';
import { Component, createSignal, ParentComponent, Show } from 'solid-js';

const Test: ParentComponent<{ renderChildren: boolean }> = p => {
  return (
    <div>
      <Show when={p.renderChildren} fallback='this should be visible only from route /1'>
        {p.children}
      </Show>
    </div>
  );
}

export const App: Component = () => {
  const [triggersBugWhenFalse, setTriggersBugWhenFalse] = createSignal(true);
  return (
    <>
      <button
        onClick={() => setTriggersBugWhenFalse(a => !a)}
      >
        {triggersBugWhenFalse() ? 'trigger bug by not rendering children of route /1' : 'render children again'}
      </button>
      <Routes>
        <Route path='' element={<Test renderChildren={triggersBugWhenFalse()}><Outlet /></Test>}>
          <Route path='1' element={<div>route 1, <NavLink href='/2'>go to 2</NavLink></div>} />
        </Route>
        <Route path='' element={<Outlet />}>
          <Route path='2' element={<div>route 2, <NavLink href='/1'>go to 1</NavLink></div>} />
        </Route>
      </Routes>
    </>
  );
};
```

Curiously, this code works as expected (change versus bugged example = the outer routes have a different path):
```
import { NavLink, Outlet, Route, Routes } from 'solid-app-router';
import { Component, createSignal, ParentComponent, Show } from 'solid-js';

const Test: ParentComponent<{ renderChildren: boolean }> = p => {
  return (
    <div>
      <Show when={p.renderChildren} fallback='this should be visible only from route /1'>
        {p.children}
      </Show>
    </div>
  );
}

export const App: Component = () => {
  const [triggersBugWhenFalse, setTriggersBugWhenFalse] = createSignal(true);
  return (
    <>
      <button
        onClick={() => setTriggersBugWhenFalse(a => !a)}
      >
        {triggersBugWhenFalse() ? 'trigger bug by not rendering children of route /1' : 'render children again'}
      </button>
      <Routes>
        <Route path='1' element={<Test renderChildren={triggersBugWhenFalse()}><Outlet /></Test>}>
          <Route path='' element={<div>route 1, <NavLink href='/2'>go to 2</NavLink></div>} />
        </Route>
        <Route path='2' element={<Outlet />}>
          <Route path='' element={<div>route 2, <NavLink href='/1'>go to 1</NavLink></div>} />
        </Route>
      </Routes>
    </>
  );
};
```

This leads me to believe conditional Outlets and the route scoring algorithm do not play nice together.

**Comment by ryansolid** [comment]
Yeah I suspect we shouldn't have(allow) the same path at the same nested level.

**Comment by toonddd** [comment]
Allowing and handling same paths at the same nesting levels opens up some imo essential routing features that are also supported by react-router v6:

Take the following routing tree:
- / (rendering wrapper layout x with an Outlet)
  - page-1
  - page-2
- / (rendering wrapper layout y with an Outlet)
  - page-3
  - page-4

Imagine layout x and y have some state, that ideally is saved when routing between page 1 <-> 2 (for wrapper x), and when routing between page 3 <-> page 4 (for wrapper y).

Even if there is no state, having to repeat the same wrapper for each leaf page is not DRY. Would be handy to group all leaf routes that should have the same wrapper in a top-level route and assign the wrapper to its element.

Are you open to a PR to make this pattern work?

**Comment by rturnq** [comment]
I see, when navigating between the two `path=''` routes, they get incorrectly identified as equivalent. In an effort to only render routes that have changed the existing one is preserved when it should not be.

We need stable identification of the author-defined routes to compare later at render time which is currently done with the route `path` and it is not good enough. I have an idea how we can keep a reference to the original route definition object and use that for equality comparisons.

**Comment by rturnq** [comment]
This is fixed in 0.4.0.
